### PR TITLE
Support other types of hash scheme to scramble the value and other scrambling enhancement

### PR DESCRIPTION
**Pull request by TejasKM**
1. Added support for other types of hash scheme to scramble the value
Currently the plugin supports MD5 as the only hashing scheme. This feature adds support for other hashing schemes. A new optional parameter hash_scheme is added to support this. Supported schemes are md5, sha1, sha256, sha384, sha512 

```
<filter **>
    @type sanitizer
    hash_salt mysalt
    hash_scheme sha512

    <rule>
        keys user
        pattern_regex /\d{3}-\d{3}/
        pattern_regex_prefix USER_ID
    </rule>

</filter>   
```

2. Added support for scrambling entire value of key
The current plugin does not support scrambling of entire value of a specified key. It requires either of pattern_ipv4, pattern_fqdn, pattern_regex, pattern_keywords to scramble value.
This feature will add support for scrambling entire value of key.

```
<filter **>
    @type sanitizer
    hash_salt mysalt
    hash_scheme sha512

   <rule>
      keys host,log
   </rule>

</filter>  
```
  
3. Added support for scrambling of multiple occurrences of pattern within the value

4. Fixed issue
In case of capture group, when the matching pattern is at end of the value, the current implementation was failing to handle this. Fixed as part the the above changes.

### CODE REVIEW

**Comment by kubotat** [comment]
@TejasKM Thanks for creating the PR. I will review it and get back to you.

**Comment by kubotat** [comment]
@TejasKM Thanks for the contribution. I've tested hash_scheme and it did work. I'm now running backward compatibility test and found discrepancies with the following configuration and input:
- Configuration
```
<system>
  log_level debug
</system>

<source>
  @type http
  port 9880
  bind 0.0.0.0
</source>

<filter **>
  @type sanitizer
  hash_salt mysalt
  hash_scheme sha512
  <rule>
    keys user.ssn
    pattern_regex /^(?!(000|666|9))\d{3}-(?!00)\d{2}-(?!0000)\d{4}$/
    pattern_regex_prefix SSN
  </rule>
  <rule>
    keys user.phone
    pattern_regex /^\d{3}-?\d{3}-?\d{4}$/
    pattern_regex_prefix Phone
  </rule>
</filter>

<match **>
  @type stdout
</match>
```
- Ingest sample data
```
# curl -X POST -d 'json={"user":{"ssn":"123-45-6789","phone":"123-456-7890"}}' http://localhost:9880/demo
```
- Output
  - There was "Skipping this key error_class=TypeError error="wrong argument type Array (expected Regexp)" which happens in "Line177 : scans = str.scan(regex)".
```
2022-02-16 06:38:44 +0000 [warn]: #0 Skipping this key error_class=TypeError error="wrong argument type Array (expected Regexp)"
2022-02-16 06:38:44 +0000 [debug]: #0 [pattern_regex] sanitize '123-456-7890' to 'Phone_0a25187902a0cf755627397eb085d736'
2022-02-16 06:38:44.084750729 +0000 demo: {"user":{"ssn":"123-45-6789","phone":"Phone_0a25187902a0cf755627397eb085d736"}}
2022-02-16 06:38:44.084894615 +0000 fluent.warn: {"error_class":"TypeError","error":"wrong argument type Array (expected Regexp)","message":"Skipping this key error_class=TypeError error=\"wrong argument type Array (expected Regexp)\""}
2022-02-16 06:38:44.085042019 +0000 fluent.debug: {"message":"[pattern_regex] sanitize '123-456-7890' to 'Phone_0a25187902a0cf755627397eb085d736'"}
```
I will look into this issue but would appreciate it if you could check this as well.

**Comment by TejasKM** [comment]
Hi @kubotat 
Thanks for taking a look at PR. I will also check from my end the error you have mentioned.

File: lib/fluent/plugin/filter_sanitizer.rb [hunk 1]
```diff
@@ -23,15 +23,18 @@ class SanitizerFilter < Fluent::Plugin::Filter
...
       helpers :event_emitter, :record_accessor
 
-      desc "Hash salt to be used to generate hash values with MD5(optional)"
+      desc "Hash salt to be used to generate hash values with specified hash(optional)"
       config_param :hash_salt, :string, default: ""
-      
-      config_section :rule, param_name: :rules, multi: true do 
+
+      desc "Hash scheme to use for generating hash value (supported schemes are md5,sha1,sha256,sha384,sha512) (optional)"
+      config_param :hash_scheme, :string, default: "md5"
```
Comment (repeatedly) [inline lib/fluent/plugin/filter_sanitizer.rb]: `enum` type is better. No need manual check.

https://docs.fluentd.org/plugin-development/api-config-types#enum

File: lib/fluent/plugin/filter_sanitizer.rb [hunk 2]
```diff
@@ -339,6 +360,22 @@ def sanitize_keywords_val(v, keywords, prefix)
         return line.join(" ")
       end
 
+      def sanitize(str)
```
Comment (repeatedly) [inline lib/fluent/plugin/filter_sanitizer.rb]: I think this method is slow. We can create callable object in `configure`

```
# Inside configure
@sanitize_func =
  case @hash_scheme
  when :md5
    Proc.new { |str| Digest::MD5.hexdigest(@salt + str) }
  when :sha1
    Proc.new { |str| Digest::SHA1.hexdigest(@salt + str) }
  # Put same code for other schemas
  end

# Inside sanitize_ipv4
return "IPv4_"+ @sanitize_func.call(str)
```

**Review by repeatedly — COMMENTED** [review]
Add comments for implementation.

**Comment by kubotat** [comment]
@repeatedly  Thank you so much for your advice. I will look into it.

File: lib/fluent/plugin/filter_sanitizer.rb (hunk 2 above)
Comment (TejasKM) [reply lib/fluent/plugin/filter_sanitizer.rb]: @repeatedly  I have made the relevant changes

**Comment by TejasKM** [comment]
@kubotat 
I checked the issue that you have mentioned. 
It is failing for pattern_regex /^(?!(000|666|9))\d{3}-(?!00)\d{2}-(?!0000)\d{4}$/

I have used the scan method to find matching pattern. Scan method returns matched data based on capture groups defined in the regex.
However, if some capture groups defined in the regex are optional and did not match, then scan returns "nil". It causes the iteration on the array to fail.

I have changed the implementation to handle this and supported the backward compatibility.

Please take a look

**Comment by TejasKM** [comment]
Hi @kubotat Can you please take a look once again

**Comment by f9-abs** [comment]
@kubotat @repeatedly can this request be approved?

**Comment by f9-abs** [comment]
@kubotat @repeatedly can we please get an approval on this pull request?

**Comment by jlangford9284** [comment]
Is there any update on this? I would like to use this feature :D

**Comment by kubotat** [comment]
Merge the PR and then fix compatibility issue later. 
Thanks for your contribution @TejasKM !!

**Pull request merged**